jsonrpc/src/cancellation.fileBased.ts

```diff
@@ -0,0 +1,93 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import * as fs from 'fs';
+import { MutableToken, AbstractCancellationTokenSource, CancellationTokenSource } from './cancellation';
+import { CancellationToken } from './cancellation';
+
+class FileBasedToken extends MutableToken {
+
+	constructor(private _cancellationName: string) {
+		super();
+	}
+
+	get isCancellationRequested(): boolean {
+		if (super.isCancellationRequested) {
+			return super.isCancellationRequested;
+		}
+
+		if (this.pipeExists()) {
+			// the first time it encounters cancellation file, it will
+			// cancel itself and raise cancellation event.
+			// in this mode, cancel() might not be called explicitly by jsonrpc layer
+			this.cancel();
+		}
+
+		return super.isCancellationRequested;
+	}
+
+	public dispose(): void {
+		super.dispose();
+
+		if (!super.isCancellationRequested) {
+			// this could be micro optimization we don't want to keep
+			return;
+		}
+
+		try {
```

Comment:
if conitnuing ("then") the promise returned from here
(https://github.com/microsoft/vscode-languageserver-node/blob/master/jsonrpc/src/main.ts#L1096) with capture of "token" and cancellation file path is okay/work, then I can go that way as well.

right now, I added in server-side since that seems more straight forward and code seems to have more clear point where I can add it. and I can't remove the possibility of leaving the cancellation file out 100% anyway, so it is in server-side for now.

Reply:
See my comments about the strategy. I think having it would allow use to experiment with this better without changing the JSON RPC code.